include/zephyr/sys/barrier.h

```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright (c) 2023 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_BARRIER_H_
+#define ZEPHYR_INCLUDE_SYS_BARRIER_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if defined(CONFIG_BARRIER_OPERATIONS_ARCH)
+/* Empty */
+#else
+#include <zephyr/sys/barrier_builtin.h>
+#endif
+
+#ifdef __cplusplus
+} /* extern "C" */
+#endif
```

Comment:
Doxygen docstrings should be present in this header, which is the only one included by API clients. Others are just implementation details.

Reply:
what about something like:
```c
#if defined(CONFIG_BARRIER_OPERATIONS_ARCH) && !defined(__DOXYGEN__)
/* Empty */
#else
#include <zephyr/sys/barrier_builtin.h>
#endif
```

Reply:
That doesn't solve the problem, I was talking about making top header the source of docstrings. Maybe the following pattern can be followed instead:

`zephyr/sys/barrier.h`:

```c
/**
 * @brief Full/sequentially-consistent data memory barrier.
 *
 * This routine acts as a synchronization fence between threads and prevents
 * re-ordering of data accesses instructions across the barrier instruction.
 */
static ALWAYS_INLINE void barrier_dmem_fence_full(void)
{
    z_barrier_dmem_fence_full();
}
```

`zephyr/sys/barrier_builtin.h` (or arch specific headers):

```c
static ALWAYS_INLINE void z_barrier_dmem_fence_full(void)
{
	__atomic_thread_fence(__ATOMIC_SEQ_CST);
}
```

We could also introduce something like this to ease unit testing (ie, an additional choice to builtin/arch) https://www.keil.com/pack/doc/CMSIS/Core/html/group__NVIC__gr.html#gadc48b4ed09386aab48fa6b9c96d9034c

Reply:
done